components/script/dom/bindings/record.rs

```diff
@@ -0,0 +1,211 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+//! The `Record` (open-ended dictionary) type.
+
+use crate::dom::bindings::conversions::jsid_to_string;
+use crate::dom::bindings::error::report_pending_exception;
+use crate::dom::bindings::str::{ByteString, DOMString, USVString};
+use indexmap::IndexMap;
+use js::conversions::{ConversionResult, FromJSValConvertible, ToJSValConvertible};
+use js::jsapi::HandleId as RawHandleId;
+use js::jsapi::JSContext;
+use js::jsapi::JS_NewPlainObject;
+use js::jsapi::PropertyDescriptor;
+use js::jsapi::JSITER_HIDDEN;
+use js::jsapi::JSITER_OWNONLY;
+use js::jsapi::JSITER_SYMBOLS;
+use js::jsapi::JSPROP_ENUMERATE;
+use js::jsval::ObjectValue;
+use js::jsval::UndefinedValue;
+use js::rust::wrappers::GetPropertyKeys;
+use js::rust::wrappers::JS_DefineUCProperty2;
+use js::rust::wrappers::JS_GetOwnPropertyDescriptorById;
+use js::rust::wrappers::JS_GetPropertyById;
+use js::rust::wrappers::JS_IdToValue;
+use js::rust::HandleId;
+use js::rust::HandleValue;
+use js::rust::IdVector;
+use js::rust::MutableHandleValue;
+use std::cmp::Eq;
+use std::hash::Hash;
+use std::marker::Sized;
+use std::ops::Deref;
+
+pub trait RecordKey: Eq + Hash + Sized {
+    fn to_utf16_vec(&self) -> Vec<u16>;
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<Self>;
+}
+
+impl RecordKey for DOMString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.encode_utf16().collect::<Vec<_>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<DOMString> {
+        jsid_to_string(cx, id)
+    }
+}
+
+impl RecordKey for USVString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.0.encode_utf16().collect::<Vec<_>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<USVString> {
+        rooted!(in(cx) let mut jsid_value = UndefinedValue());
+        let raw_id: RawHandleId = id.into();
+        JS_IdToValue(cx, *raw_id.ptr, jsid_value.handle_mut());
+
+        match USVString::from_jsval(cx, jsid_value.handle(), ()) {
+            Ok(ConversionResult::Success(s)) => Some(s),
+            Ok(ConversionResult::Failure(_)) => None,
+            Err(..) => None,
+        }
+    }
+}
+
+impl RecordKey for ByteString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.iter().map(|&x| x as u16).collect::<Vec<u16>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<ByteString> {
+        rooted!(in(cx) let mut jsid_value = UndefinedValue());
+        let raw_id: RawHandleId = id.into();
+        JS_IdToValue(cx, *raw_id.ptr, jsid_value.handle_mut());
+
+        match ByteString::from_jsval(cx, jsid_value.handle(), ()) {
+            Ok(ConversionResult::Success(s)) => Some(s),
+            Ok(ConversionResult::Failure(_)) => None,
+            Err(..) => None,
+        }
+    }
+}
+
+/// The `Record` (open-ended dictionary) type.
+#[derive(Clone, JSTraceable)]
+pub struct Record<K: RecordKey, V> {
```

Comment:
Putting trait bounds on type definitions often make other code more verbose for no good reason, I would recommend removing it from here and other places where it's not required.

Reply:
It seems it's needed for `JSTraceable`. Strings are traceable but K is not 🤔

Reply:
I'm confused here, all `RecordKey` does is extend `Eq + Hash + Sized`, none of those are related to `JSTraceable`. Could you show me what is the error without that trait bound?

Reply:
Yup, the error:

```
error[E0599]: no method named `trace` found for type `&indexmap::map::IndexMap<K, V>` in the current scope
  --> components\script\dom\bindings\record.rs:88:17
   |
88 | #[derive(Clone, JSTraceable)]
   |                 ^^^^^^^^^^^ method not found in `&indexmap::map::IndexMap<K, V>`
   |
   = note: the method `trace` exists but the following trait bounds were not satisfied:
           `indexmap::map::IndexMap<K, V> : dom::bindings::trace::JSTraceable`
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `trace`, perhaps you need to implement it:
           candidate #1: `dom::bindings::trace::JSTraceable`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `script`.
```

Reply:
Oh I see!

This is because of the way the `JSTraceable` impl for `IndexMap<K, V>` is written:

https://github.com/servo/servo/blob/db1750673c81de852a3bfd6ad79941f4622eab94/components/script/dom/bindings/trace.rs#L372-L385

If you remove `Eq + Hash` there, it should work without the `RecordKey` bound here.

Reply:
It kinda works, and then it fails because of `IntoIterator` impl for `IndexMap`:

```rust
impl<'a, K, V, S> IntoIterator for &'a IndexMap<K, V, S>
    where K: Hash + Eq,
          S: BuildHasher,
{
    type Item = (&'a K, &'a V);
    type IntoIter = Iter<'a, K, V>;
    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}
```

```
error[E0277]: the trait bound `&indexmap::map::IndexMap<K, V, S>: std::iter::IntoIterator` is not satisfied
   --> components\script\dom\bindings\trace.rs:380:23
    |
380 |         for (k, v) in &*self {
    |                       ^^^^^^ the trait `std::iter::IntoIterator` is not implemented for `&indexmap::map::IndexMap<K, V, S>`
    |
    = help: consider adding a `where &indexmap::map::IndexMap<K, V, S>: std::iter::IntoIterator` bound
    = note: required by `std::iter::IntoIterator::into_iter`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `script`.
```

Reply:
It should work if you do `for (k, v) in self.iter() { … }`. For some reason the `IntoIterator` impl has those bounds but not the inherent `IndexMap::iter` method.

Reply:
Seems it also has the bounds?

```
error[E0599]: no method named `iter` found for type `&indexmap::map::IndexMap<K, V, S>` in the current scope
   --> components\script\dom\bindings\trace.rs:380:28
    |
380 |         for (k, v) in self.iter() {
    |                            ^^^^ method not found in `&indexmap::map::IndexMap<K, V, S>`
    |
    = note: the method `iter` exists but the following trait bounds were not satisfied:
            `K : std::cmp::Eq`
            `K : std::hash::Hash`
```

Reply:
Ok guess I can't read code anymore, disregard all that and sorry for the wasted time.

Reply:
Your suggestion looked good, too bad we can't apply it 😭

Reply:
(Hmm, I totally can't remember why I implemented `JSTraceable` for `IndexMap` with `Hash + Eq` for the type `K` 😭 )

Reply:
Because you need those to iterate, don't worry about that. :)

Reply:
@nox Ah, I see! Related to @saschanaz 's https://github.com/servo/servo/pull/24377#discussion_r332494388 ! Understood and thanks :D